python/cudf/cudf/core/spillable_buffer.py

```diff
@@ -0,0 +1,445 @@
+            if self.spillable:
+                self.__spill__(target="cpu")
+                return self._ptr_desc["memoryview"][offset : offset + size]
+            else:
+                assert self._ptr_desc["type"] == "gpu"
+                ret = memoryview(bytearray(size))
+                rmm._lib.device_buffer.copy_ptr_to_host(
+                    self._ptr + offset, ret
+                )
+                return ret
+
+    def getitem(self, offset: int, size: int) -> Buffer:
+        return SpillableBufferView(base=self, offset=offset, size=size)
+
+    def serialize(self) -> Tuple[dict, List[Frame]]:
+        header: Dict[Any, Any]
+        frames: List[Frame]
+        with self._lock:
+            header = {}
+            header["type-serialized"] = pickle.dumps(self.__class__)
+            header["frame_count"] = 1
+            if self.is_spilled:
+                frames = [self.memoryview()]
+            else:
+                spill_lock = SpillLock()
+                ptr = self.get_ptr(spill_lock=spill_lock)
+                frames = [
+                    Buffer(
+                        data=ptr,
+                        size=self.size,
+                        owner=(self._owner, spill_lock),
+                    )
+                ]
+            return header, frames
+
+    @classmethod
+    def deserialize(cls, header: dict, frames: list) -> SpillableBuffer:
+        from cudf.core.spill_manager import global_manager
+
+        if header["frame_count"] != 1:
+            raise ValueError(
+                "Deserializing a SpillableBuffer expect a single frame"
+            )
+        (frame,) = frames
+        if isinstance(frame, SpillableBuffer):
+            ret = frame
+        else:
+            ret = cls(frame, exposed=False, manager=global_manager.get())
+        return ret
+
+    def is_overlapping(self, ptr: int, size: int):
+        with self._lock:
+            return (
+                not self.is_spilled
+                and (ptr + size) > self._ptr
+                and (self._ptr + self._size) > ptr
+            )
+
+    def __repr__(self) -> str:
+        if self._ptr_desc["type"] != "gpu":
+            ptr_info = str(self._ptr_desc)
+        else:
+            ptr_info = str(hex(self._ptr))
+        return (
+            f"<SpillableBuffer size={format_bytes(self._size)} "
+            f"spillable={self.spillable} exposed={self.exposed} "
+            f"num-spill-locks={len(self._spill_locks)} "
+            f"ptr={ptr_info} owner={repr(self._owner)}"
+        )
+
+
+class SpillableBufferView(SpillableBuffer):
+    """A view of a spillable buffer
+
+    This buffer deligates must operations to its base buffer.
+
+    Parameters
+    ----------
+    base : SpillableBuffer
+        The base of the view
+    offset : int
+        Memory offset into the base buffer
+    size : int
+        Size of the view (in bytes)
+    """
+
+    def __init__(self, base: SpillableBuffer, offset: int, size: int) -> None:
```

Comment:
I'm surprised that mypy isn't complaining that this constructor isn't compatible with those of the parent classes. In this case it wouldn't just be a mypy issue: the ways in which the different types of `Buffer` differ seem quite odd and potentially problematic, but I'm not quite sure yet how to unify them more tightly. I'll think on this a bit more for the next pass of review.

Reply:
Generally, I would think that it is ok to have different constructor signatures since the caller specifies the class type explicitly at creation?